returnn/tf/layers/basic.py

```diff
@@ -2685,6 +2685,83 @@ def get_out_data_from_opts(cls, name, axis, dim=1, sources=(), **kwargs):
+    else:
+      input_data = input_data.copy_as_batch_major().copy_move_axis(original_axis, 1)
+    repetitions_data = repetitions_data.copy_as_batch_major()
+    # pad the target axis
+    paddings = [(0, 1) if i == 1 else (0, 0) for i in range(input_data.batch_ndim)]
+    padded_data = tf.pad(input_data.placeholder, paddings)  # [B, T+1, ...]
+    # those are the sequence lengths after expansion
+    target_seq_len = tf.reduce_sum(repetitions_data.placeholder, axis=1)  # [B], the new size_placeholder for 'T'
+    # maximum sequence length after expansion
+    max_duration = tf.reduce_max(target_seq_len)  # [] == T'
+    # the new padding is the difference of the maximum to the new target
+    target_padding_steps = tf.expand_dims(max_duration - target_seq_len, 1)  # [B, 1]
+    # add the repetitions for the artificial padding position
+    target_repetitions = tf.concat([repetitions_data.placeholder, target_padding_steps], axis=1)  # [B, T+1]
+    # flatten batch and time
+    shape = tf.shape(padded_data)
+    flat_shape = tf.concat([[shape[0] * shape[1]], shape[2:]], axis=0)
+    reshaped_data = tf.reshape(padded_data, flat_shape)  # [B * T+1, F]
+    reshaped_repetitions = tf.reshape(target_repetitions, (shape[0] * shape[1],))  # [B * T+1]
+    # run the repetition
+    repeated_data = tf.repeat(reshaped_data, reshaped_repetitions, axis=0)  # [B * T', ...]
+    # unflatten the output
+    target_shape = tf.concat([[shape[0], max_duration], shape[2:]], axis=0)
+    self.output.placeholder = tf.reshape(repeated_data, target_shape)  # [B, T', ...]
+    # set the size_placeholder
+    self.output.size_placeholder = self.input_data.size_placeholder
+    self.output.size_placeholder[0] = target_seq_len
+
+  @classmethod
+  def get_out_data_from_opts(cls, name, axis, repetitions, sources=(), **kwargs):
+    """
+    :param str name:
+    :param str axis:
+    :param LayerBase repetitions:
+    :param list[LayerBase] sources:
+    :rtype: Data
+    """
+    data = get_concat_sources_data_template(sources)
+    data.name = "%s_output" % name
+    repetitions_data = repetitions.output.copy_template()
+    axis = data.get_axis_from_description(axis)
+    axis_wo_batch = data.get_batch_axis_excluding_batch(axis)
+    if axis < data.batch_dim_axis:
+      data = data.copy_as_batch_major().copy_move_axis(axis+1, 1)
+    else:
+      data = data.copy_as_batch_major().copy_move_axis(axis, 1)
+    assert repetitions.output.dtype in ["int32", "int64"]
+    assert repetitions_data.have_batch_axis() and repetitions_data.batch_ndim == 2, (
+      "RepeatLayer only accepts repetitions of the shape (B, None|S) or (None|S, B)")
+    repetitions_data = repetitions_data.copy_as_batch_major()
+    assert data.shape[0] == repetitions_data.shape[0], (
+      "Axis mismatch between input (%i) and repetitions (%i)" %
+      (data.shape[axis_wo_batch], repetitions_data.shape[0]))
+    return data
```

Comment:
I wonder, this would not be batch-major, but use the input data as-is. This is wrong, right?
(However, with my suggested code change above, it would be correct.)